src/twisted/protocols/_smb/tests.py

```diff
@@ -0,0 +1,363 @@
+        with self.assertRaises(_base.SMBError):
+            manager.receiveToken(
+                b"NTLMSSP\x00\xFF\0\0\0invalid message"
+                + b"type                             "
+            )
+
+
+@implementer(IFilesystem)
+class TestDisc:
+    pass
+
+
+@implementer(ISMBServer)
+class TestAvatar:
+    def getShare(self, name: str) -> Union[IFilesystem, IPipe, IPrinter]:
+        if name == "share":
+            return TestDisc()
+        else:
+            raise NoSuchShare(name)
+
+    def listShares(self) -> List[str]:
+        return ["share"]
+
+    session_id: int = 0
+
+
+@implementer(portal.IRealm)
+class TestRealm:
+    def requestAvatar(
+        self, avatarId: str, mind: core.SMBMind, *interfaces: Any
+    ) -> Tuple[type, ISMBServer, Callable[[], None]]:
+        log.debug("avatarId={a!r} mind={m!r}", a=avatarId, m=mind)
+        return (ISMBServer, TestAvatar(), lambda: None)
+
+
+class ChatNotFinished(Exception):
+    pass
+
+
+ChatType = List[Tuple[str, Optional[str]]]
+
+
+class ChatProcess(ProcessProtocol):
+    def __init__(self, chat: ChatType, ignoreRCode: bool) -> None:
+        self.chat = chat
+        self.d: Deferred = Deferred()
+        self.matches: List[Match] = []
+        self.ignoreRCode = ignoreRCode
+
+    def outReceived(self, bdata: bytes) -> None:
+        data = bdata.decode("utf-8")
+        if self.chat:
+            prompt, reply = self.chat[0]
+            m = re.search(prompt, data)
+            if m:
+                self.matches.append(m)
+                if reply:
+                    for i in range(1, 10):
+                        t = "\\%d" % i
+                        if t in reply:
+                            reply = reply.replace(t, m.group(i))
+                    if self.transport is not None:
+                        self.transport.write(reply.encode("utf-8"))
+                else:
+                    cast(IProcessTransport, self.transport).closeStdin()
+                del self.chat[0]
+
+    def errReceived(self, data: bytes) -> None:
+        log.debug("STDERR: {data!r}", data=data)
+
+    def processEnded(self, status: Failure) -> None:
+        if (not self.ignoreRCode) and status.value.exitCode != 0:
+            self.d.errback(status)
+        elif self.chat:
+            try:
+                raise ChatNotFinished()
+            except BaseException:
+                self.d.errback(Failure())
+        else:
+            self.d.callback(self.matches)
+
+
+def spawn(
+    chat: ChatType, args: Sequence[str], ignoreRCode: bool = False, usePTY: bool = True
+) -> Deferred:
+    pro = ChatProcess(chat, ignoreRCode)
+    cast(IReactorProcess, reactor).spawnProcess(pro, args[0], args, usePTY=usePTY)  # type: ignore
+    return pro.d
+
+
+TESTPORT = 5445
+SMBCLIENT = "/usr/bin/smbclient"
+
+
+@python_unittest.skipUnless(os.access(SMBCLIENT, os.X_OK), "smbclient unavailable")
+class SambaClientTests(unittest.TestCase):
```

Comment:
Including the word "integration" here should help future maintainers know that this is going to be slow and potentially flaky, which is useful to understand in advance.  It's great to have this test, just want to make sure that it is appropriately categorized and most narrow unit-tests of behavior should be tested elsewhere as well.

```suggestion
class SambaClientIntegrationTests(unittest.TestCase):
```